### PR TITLE
backport z-level draw

### PR DESCRIPTION
**Pull request by sonphantrung**
#### Summary
None

#### Purpose of change
I want to see zeds from below

#### Describe the solution
Port the change from [BN](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1743) by @joveeater

#### Describe alternatives you've considered
Not doing this

#### Testing
TBD, I don't have a good machine to compile

#### Additional context
my head is currently sleepy, I'll keep an eye on the code later ;)

### CODE REVIEW

**Comment by I2pRandom** [comment]
It took a while to bring Z-levels to BN in the first place, i don't recommend not testing this. it had a few bugs and problems along the way of getting it working.

**Comment by Lamandus** [comment]
yes, I would consider to wait just a few days until the bugs are ironed out, for example the thing with the cardboard box. It should also be tested with DDA since your code is different in some places.

**Comment by sonphantrung** [comment]
@jbytheway How do I deal with ["construction of (tri)point can be simplified"](https://github.com/CleverRaven/Cataclysm-DDA/pull/61074/files#diff-c8561880131eb6314d85f58032f9063284bcfc226b75b3b5df585dcaa219605cR1379)? I'm not a C++ professional

**Comment by scarf005** [comment]
many of the warnings are from usage of `{temp.x, temp.y, center.z}`.
since there's a variable `pos` defined as `const tripoint pos( temp, center.z );`,
you could replace them with `pos`.

- check [`struct tripoint` at src/point.h](https://github.com/CleverRaven/Cataclysm-DDA/blob/89803505c4ed1651a01169f1e8c207a4e86e8b87/src/point.h#L142) for more constructor options.
- check `CHECK-FIXES` in [clang-tidy plugin for simplify point constructor](https://github.com/CleverRaven/Cataclysm-DDA/blob/89803505c4ed1651a01169f1e8c207a4e86e8b87/tools/clang-tidy-plugin/test/simplify-point-constructors.cpp) for available fixes.

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -1390,9 +1390,9 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
             // Add scent value to the overlay_strings list for every visible tile when
             // displaying scent
             if( g->display_overlay_state( ACTION_DISPLAY_SCENT ) && !invis ) {
-                const int scent_value = get_scent().get( {temp.x, temp.y, center.z} );
+                const int scent_value = get_scent().get( pos );
                 if( scent_value > 0 ) {
-                    overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) ) + half_tile,
+                    overlay_strings.emplace( player_to_screen( point( temp ) ) + half_tile,
```
Comment (scarf005) [inline src/cata_tiles.cpp:1395]:
```suggestion
                    overlay_strings.emplace( player_to_screen( temp ) + half_tile,
```
`temp` is already of type `point`, so you don't need to wrap it

File: src/cata_tiles.cpp [hunk 2]
```diff
@@ -1401,26 +1401,26 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
             // Add scent type to the overlay_strings list for every visible tile when
             // displaying scent
             if( g->display_overlay_state( ACTION_DISPLAY_SCENT_TYPE ) && !invis ) {
-                const scenttype_id scent_type = get_scent().get_type( {temp.x, temp.y, center.z} );
+                const scenttype_id scent_type = get_scent().get_type( pos );
                 if( !scent_type.is_empty() ) {
-                    overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) ) + half_tile,
+                    overlay_strings.emplace( player_to_screen( point( temp ) ) + half_tile,
```
Comment (scarf005) [inline src/cata_tiles.cpp:1406]:
```suggestion
                    overlay_strings.emplace( player_to_screen( temp ) + half_tile,
```

File: src/cata_tiles.cpp [hunk 3]
```diff
@@ -1401,26 +1401,26 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
...
-                    const int rad_value = rad_overridden ? rad_override->second : here.get_radiation( {temp.x, temp.y, center.z} );
+                    const int rad_value = rad_overridden ? rad_override->second : here.get_radiation( pos );
                     catacurses::base_color col;
                     if( rad_value > 0 ) {
                         col = catacurses::green;
                     } else {
                         col = catacurses::cyan;
                     }
-                    overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) ) + half_tile,
+                    overlay_strings.emplace( player_to_screen( point( temp ) ) + half_tile,
```
Comment (scarf005) [inline src/cata_tiles.cpp:1423]:
```suggestion
                    overlay_strings.emplace( player_to_screen( temp ) + half_tile,
```

File: src/cata_tiles.cpp [hunk 4]
```diff
@@ -1437,7 +1437,7 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
                     } else {
                         color = catacurses::white;
                     }
-                    overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) + half_tile,
+                    overlay_strings.emplace( player_to_screen( point( temp ) + half_tile,
```
Comment (scarf005) [inline src/cata_tiles.cpp:1440]:
```suggestion
                    overlay_strings.emplace( player_to_screen( temp + half_tile,
```

File: src/cata_tiles.cpp [hunk 5]
```diff
@@ -1470,25 +1470,25 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
                     temp_str = std::to_string( units::to_kelvin( temp_value ) );
 
                 }
-                overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) ) + half_tile,
+                overlay_strings.emplace( player_to_screen( point( temp ) ) + half_tile,
```
Comment (scarf005) [inline src/cata_tiles.cpp:1473]:
```suggestion
                overlay_strings.emplace( player_to_screen( temp ) + half_tile,
```

File: src/cata_tiles.cpp [hunk 6]
```diff
@@ -1470,25 +1470,25 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
...
-                const bool visibility = g->displaying_visibility_creature->sees( {temp.x, temp.y, center.z} );
+                const bool visibility = g->displaying_visibility_creature->sees( pos );
 
                 // color overlay.
                 SDL_Color block_color = visibility ? windowsPalette[catacurses::green] :
                                         SDL_Color{ 192, 192, 192, 255 };
                 block_color.a = 100;
                 color_blocks.first = SDL_BLENDMODE_BLEND;
-                color_blocks.second.emplace( player_to_screen( point( temp.x, temp.y ) ), block_color );
+                color_blocks.second.emplace( player_to_screen( point( temp ) ), block_color );
```
Comment (scarf005) [inline src/cata_tiles.cpp:1487]:
```suggestion
                color_blocks.second.emplace( player_to_screen( temp ), block_color );
```

File: src/cata_tiles.cpp [hunk 7]
```diff
@@ -1470,25 +1470,25 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
...
 
                 // overlay string
                 std::string visibility_str = visibility ? "+" : "-";
-                overlay_strings.emplace( player_to_screen( point( temp.x, temp.y ) + quarter_tile ),
+                overlay_strings.emplace( player_to_screen( point( temp ) + quarter_tile ),
```
Comment (scarf005) [inline src/cata_tiles.cpp:1491]:
```suggestion
                overlay_strings.emplace( player_to_screen( temp + quarter_tile ),
```

File: src/cata_tiles.cpp [hunk 8]
```diff
@@ -1501,7 +1501,7 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
                     SDL_Color blue = { 0, 0, 255, 255 };
                     lighting_colors = color_linear_interpolate( white, blue, 9 );
                 }
-                point tile_pos = player_to_screen( point( temp.x, temp.y ) );
+                point tile_pos = player_to_screen( point( temp ) );
```
Comment (scarf005) [inline src/cata_tiles.cpp:1504]:
```suggestion
                point tile_pos = player_to_screen( temp );
```

**Review by scarf005 — CHANGES_REQUESTED** [review]
`point ( temp )` can be furthur simplified to `temp`, since type of `temp` is `point`.
https://github.com/CleverRaven/Cataclysm-DDA/blob/5a592543b7a34c868ac33d78c0479ab7293f2d66/src/cata_tiles.cpp#L1362-L1375

**Comment by kevingranade** [comment]
This needs feature parity with curses, I also need to see what it does to memory requirements for tilesets as it essentially makes one (or more?!?) copies of the tileset in use.

**Comment by sonphantrung** [comment]
Oh yeah, I wish I could test this, but I currently don’t have something good to compile, and I’m fairly busy now. Shall we discuss?

**Comment by kevingranade** [comment]
I'm not clear what there is to discuss. There's certain criteria that I briefly outlined that need to be met and that either gets met or it doesn't.

**Comment by Fris0uman** [comment]
Porting the feature will probably require a fair bit of work on one of the most obscure part of the codebase, are you sure you're able to do that? Might be better to see what RAldrich comes up with since they were apparently intereted in fixing #58479

**Comment by sonphantrung** [comment]
I'm currently having some pretty ugly merge conflict problems, plus seems like the code is pretty messed up since the PR initially opened, so I'll close this for the better second attempt, when my head isn't sleepy ;)

**Comment by RAldrich** [comment]
Bit of thread necromancy here, but I wanted to say not to wait on me to do anything with #58479. That's on hold indefinitely; getting it to an acceptable state of polish requires large changes to the rendering engine itself, which (like FrisOuman called out) is an extremely major project. For some perspective, I'm a sr. software engineer and I haven't been able to fully wrap my head around how it currently works (much less how it **should** work). Maybe someday when I have more time!

**Pull request closed without merging**